app/api/views.py

```diff
@@ -0,0 +1,224 @@
+from app.prompts.models import Prompt
+from app.responses.models import PromptResponse, Response
+
+
+class ValidateFeedbackFormMixin:
+    def validate_feedback_form_exists(
+        self, url_params: dict[str, str], ensure_enabled=False
+    ):
+        """
+        Validates the FeedbackForm exists (and optionally is enabled) in the Project
+        """
+
+        feedback_form_query = FeedbackForm.objects.filter(
+            uuid=url_params["feedback_form_id"],
+            project__uuid=url_params["project_id"],
+        )
+
+        if ensure_enabled:
+            feedback_form_query = feedback_form_query.filter(disabled_at=None)
+
+        if not feedback_form_query.exists():
+            raise NotFound(
+                " ".join(
+                    [
+                        f"Feedback form id={url_params["feedback_form_id"]} does not exist",
+                        *(["or is disabled"] if ensure_enabled else []),
+                        f"in project id={url_params["project_id"]}.",
+                    ]
+                )
+            )
+
+
+class ValidateResponseMixin:
+    def validate_response_exists(self, url_params: dict[str, str]):
+        """
+        Validates the Response exists for the FeedbackForm
+        """
+
+        response_exists = Response.objects.filter(
+            uuid=url_params["response_id"],
+            feedback_form__uuid=url_params["feedback_form_id"],
+        ).exists()
+
+        if not response_exists:
+            raise NotFound(
+                f"Response id={url_params["response_id"]} does not exist in feedback form id={url_params["feedback_form_id"]}."
+            )
+
+
+class FeedbackFormDetail(generics.RetrieveAPIView):
+    queryset = FeedbackForm.objects.all()
+    serializer_class = FeedbackFormSerializer
+
+    def get_queryset(self):
+        return self.queryset.filter(
+            uuid=self.kwargs["feedback_form_id"],
+            project__uuid=self.kwargs["project_id"],
+        ).prefetch_related(
+            Prefetch(
+                "prompts",
+                queryset=Prompt.objects.select_subclasses(),
+            ),
+        )
+
+    def get_object(self):
+        return generics.get_object_or_404(self.get_queryset())
+
+
+class PromptList(generics.ListAPIView, ValidateFeedbackFormMixin):
+    queryset = Prompt.objects.all()
+    serializer_class = PromptSerializer
+
+    def get_queryset(self):
+        return (
+            self.queryset.filter(
+                feedback_form__uuid=self.kwargs["feedback_form_id"],
+            )
+            .order_by("order")
+            .select_subclasses()
+        )
+
+    def get(self, request, *args, **kwargs):
+        self.validate_feedback_form_exists(kwargs)
+        return super().get(request, *args, **kwargs)
+
+
+class ResponseListCreate(generics.ListCreateAPIView, ValidateFeedbackFormMixin):
+    queryset = Response.objects.all()
+    serializer_class = ResponseSerializer
```

Comment:
Is there a risk that using the same serializer class for both reading and writing can inadvertently leak data on creation that we don't want to leak?

Reply:
I'll have a look.

Reply:
We covered this on a call today. I think it's generally about code maintainability & accidental complexity & element of surprise vs code duplication. And also if we even want to allow reading data now.